### DISCUSSION TITLE
Multi Language content update via the Rest API

### DISCUSSION THREAD

**Discussion by rockgobo** [2sic/2sxc#2362] [Q&A] [answered]
Is it possible to update the translation of a content item via the Rest API? I tried to post the translation to the second language endpoint (like .../en-us/api/2sxc/app/Content/content/projects) but nothing changed. The translation still sticks to the first language as fallback. 

I would like to write a little piece of software that will live in the scheduler of dnn and updates the content in 2sxc from a external API. So i need to create items (which works fine so far), updates the item (which is also possible) and then it would do the same procedure for the second language of the website (this is not working right now). 

Does anybody have the some use case?

**Comment by iJungleboy** [comment, marked as answer]
Hi @rockgobo 

As of now the API is still only simple - that means single language only. To do multi-language data it would require quite a bit of skill in the 2sxc / EAV internal C# APIs.

If you need this and would be willing to invest some time with me, then I would extend the API but would need you to do testing etc. till it works. I'll probably need 2-4 hours and you as well. Are you in?

**Reply by rockgobo** [reply]
Sure, that would be awesome! I am definitely in.

**Comment by iJungleboy** [comment]
So let's start specifying what we need. I assume you have data like 

```
{
  "Name": "Something",
  "Message": "Something"
}
```

Now you need to create such data in multi-language. Can this be done in one call or do the calls need to be separated (like first primary language, a while later the secondary language? This will affect if we have a new format with multiple values or if we need to implement partial-updates.

**Reply by rockgobo** [reply]
That is a realy good question. If all non primary languages would have a specific route, it would feel a little bit more natural and flexible to me, but i guess it would mean that the first call to create a content item always have to target the primary language. Maybe also that you cannot create data on the secondary language route:

```
POST \projects creates a new project
POST \projects\id updates the project
POST \projects\id\en-us updates the secondary language
```

Having separate routes would mean the payload of the message has always the some structure and you don´t have to write a language coded format, like

```
{
  "en-us": {
    "Name": "Something",
    "Message": "Something"
  },
  "de-de": {
    "Name": "Irgendein Name",
    "Message": "Eine Nachricht"
  }
}
```

As both scenarios would work in my specific usecase, what is your favorite way of doing it?

**Reply by iJungleboy** [reply]
Since the 2sxc/EAV data model is each value can (but doesn't have to) be multi-language, I think it's important we stay consistent with that concept. Otherwise docs and APIs get really confusing.

I believe it should also have some kind of version-header, so future API changes won't break existing code. My recommendation below is a slimmed down version of the official internal JSON format https://docs.2sxc.org/basics/data-formats/json-v01/value.html

1. I would also like to add a `$schema` json header to identify the format and make it easier for people who see an example to find docs about it
2. And I think there should be a simple way to set values which are not translated / just saved on the primary language
3. The optional "_" node would also contain header information like [Metadata-Keys](https://docs.2sxc.org/basics/data-formats/json-v01/metadata.html)
4. We should also anticipate that [files may be included directly](xref:https://docs.2sxc.org/basics/data-formats/json-v01/assets.html) (probably not in the first implementation) but I would like to plan ahead...

So my guess is something like this:

```json
{
  "$schema": "https://docs.2sxc.org/schemas/json/save-entity2021",
  "_": {
    "For": {
      "Target": "CmsObject",
      "String": "file:72"
    }
  },
  "Name": {
    "en-us": "Something",
    "de-de": "Irgendein Name"
  },
  "Message": {
    "en-us": "Good Day",
    "de-de": "Guten Tag"
  },
  "Rating": 7,
  "Source": "iOS 14 Native App",
  "Tags": [ 
    "dae148ec-039b-41dd-9249-fba21487c32a", 
    "3f5d5c3e-9f04-4cad-9eab-b0a4e36adddc",
    "c4db1bc0-41a3-44fe-a6ea-006c9354627f"
  ],
}
```

Or for Metadata it would be a bit like this:

```json
{
  "$schema": "https://docs.2sxc.org/schemas/json/save-entity2021",
  "_": {
    "For": {
      "Target": "CmsObject",
      "String": "file:72"
    }
  },
  "NSFW": true,
  "Description": { 
      "en-us": "iOS 14 Native App"
  },
}
```

Or with attachments (future implementation) a bit like this:

```json
{
  "$schema": "https://docs.2sxc.org/schemas/json/save-entity2021",
  "Title": "Homepage of 2sxc",
  "Logo": "http://xyz",
  "Screenshot": {
      "Assets": [
        {
        "Name": "logo.gif",
        "File": "R0lGODdhzABuAPcAAAAAADMAAGYAAJkAAMwAAP8AAAAzADMzAGYzAJkzAMwzAP8zAABmADNmAGZmAJlmAMxmAP9mAACZADOZAGaZAJmZAMyZAP+ZAADMADPMAGbMAJnMAMzMAP/MAAD/ADP/AGb/AJn/AMz/AP//AAAAMzMAM2YAM5kAM8wAM/8AMwAzMzMzM2YzM5kzM8wzM/8zMwBmMzNmM2ZmM5lmM8xmM/9mMwCZMzOZM2aZM5mZM8yZM/+ZMwDMMzPMM2bMM5nMM8zMM//MMwD/MzP/M2b/M5n/M8z/M///MwAAZjMAZmYAZpkAZswAZv8AZgAzZjMzZmYzZpkzZswzZv8zZgBmZjNmZmZmZplmZsxmZv9mZgCZZjOZZmaZZpmZZsyZZv+ZZgDMZjPMZmbMZpnMZszMZv/MZgD/ZjP/Zmb/Zpn/Zsz/Zv//ZgAAmTMAmWYAmZkAmcwAmf8AmQAzmTMzmWYzmZkzmcwzmf8zmQBmmTNmmWZmmZlmmcxmmf9mmQCZmTOZmWaZmZmZmcyZmf+ZmQDMmTPMmWbMmZnMmczMmf/MmQD/mTP/mWb/mZn/mcz/mf//mQAAzDMAzGYAzJkAzMwAzP8AzAAzzDMzzGYzzJkzzMwzzP8zzABmzDNmzGZmzJlmzMxmzP9mzACZzDOZzGaZzJmZzMyZzP+ZzADMzDPMzGbMzJnMzMzMzP/MzAD/zDP/zGb/zJn/zMz/zP//zAAA/zMA/2YA/5kA/8wA//8A/wAz/zMz/2Yz/5kz/8wz//8z/wBm/zNm/2Zm/5lm/8xm//9m/wCZ/zOZ/2aZ/5mZ/8yZ//+Z/wDM/zPM/2bM/5nM/8zM///M/wD//zP//2b//5n//8z//////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAAAAAAALAAAAADMAG4AQAj/AF0IHEiwoMGDCBMqXMiwocOHECNKnEixosWLGDNq3Mixo8ePID1eG0mypMmSIVMaPMny2peBWFqZbNUqUSKaJgd9THSSZs2bOGe+pNiy6EiVSAUeTeqCZyuHMp+qHDRSp0KjRZmqXHr1mtWFMr8elDl05bUsCaNWVItQJta3cN9qbcj1IM9EEMmmdWnwyzWpcw1iiUu4MMnAC+sa1PuQ8Vi+BqkCjqxYq+HKgalew5JwJN7O1xQ6bji6YGmzkxUOCkqS5iDODP3+BW0YMWWvt1Obtcm7N0/IUP/67j2yLETNI1uJXZhldWvjdC/bnk69uvXrBlesOKg9+3aELb4T/1zBwvvD7gRZaEc/XnzB8C0Kkjffnrt7F+HLY9/P3zb7/gAGKOCABBZo4IEIJqjggtVdFhqDB2E2kFuzUeTWcg5l4dZnGMWU3EMOQhjhgwxlQVIisAmkoWfMSSiQZmAlB11fbulGEHIO5pgVfy664JeNoIkGHEEjoVUdhTom2SNiPZ4m5IwTDinQYCQa9JuSWGYJl0dXnuhQk1LGCKVATrpAZYqP4eaQh3+VxCFDITZYJUI9gjlmW2FGeadpH350IZzSiSjooITWRx926hU6KHzvrRBfQv+5kGhDkS5UqQuVXoqfowTlZ9CkA10aKaOKlmrqqaimquqqrLbq6quwxv8qa0RxzmornYHealAWXwxi0yB7hoTFF1/0RqyRGtVKqJ0myujQlRhKhJxyEq2YHJrR1aYomMhaxFaGfXqkGZAFKTvokpEheS1paiYkW7QEsYlVK9guViGu2haKrgvf4kmuUnm6AGNCv3XbkGZvFoSjljn2t2RYDv34pF332qsluQwn6fCcBd0FkZ277quSbBnnih2Y8OJ552kSX1eyudaB3FjA/OZZHEJUUgeziDIHt3LAVSWkmcEPUZVyV/lSR2W7ArtFdLkcE1SmynvdSWWwFNb0BZrFZo2tc+vSVljMZyUW9UA9s/tzsAIHrZLTSI99ZMX2/ps2Q1NPfRByCVOhJO/TIyZd3cJeMjTc4Sgeh7hv9UaEBdhxUavr5JRXJ+p916nHQguct6Bf5UldblB+nh8UnqOez5cdC5ozBCqlrKveqXacq/eofPeJyvrmp99eqO6ma/d5qOsVnynmCb1u6PK4H0p8884LpCmEvXNOO3eb2z668J0P/zxEp/OeOubTV3+9fbXHbh/r1nMK+vvQwy///PTXb//9+Oev//789+////8ADCCAdiZABhGwgAo6IAIRpMAFGqiBCSKWBCdIwb2pSygqycJN3uIajECQgX/xiQiDMpQrKadxPvLaRrp0woWsJmtsC5zc9HU2MrXJbQ9h09HwRpIYKgRaHzPZuWpYM6ZNBG4QaVbZOPKbfxFJiILqUSIAJxEq7fCJTrxIs7L4wQOJbFfEQiGuxBilzRzsgn+5YtuWiK8ZLouIBbFWcpB0RcnEyInTAlYcIbfDl2mMR3BUUdhuE6w6XaNvUsMhQ8YVNz+60WWB/E10qFizMfUrjoos2s3a6Mi4bMyFIrNj1fbmItmQcSFUAhzhOmmUT3bllDC5mw3hpUQt7YWV/1sCpGroZraJmeU6q8TlSVyJEI/NbG194SViSCbMHe1HiofMC81YpkwmNbOVuhzlMbVJkJYdZFgUDKc4x0lOCV4Tm88kot7shcxfImRcI4ynPON5Th0R8zE+tGE7+ZQyiNmmiw9UJ83wyU2CNKsrahQJFK0DTlg+0Zc+K2gi/4VEaZIkoWhbKGLkyBo2BkmiYgJplJxYo1PKyyo4ohegBGebS5axlyKNaSWFFEh5GaUVd/pC1sRGGOvIBJEDESWuIKo2mQoymklxSyM9KSdAdYWoPNwnTTF6xEyahaWB8SdCDvpRqm3Tqw1pokMfQjI1AhQkVnznJodqVHa2VTAUAv1qQ+SVxSdiFTEUAlavarRSrOSziH6tCOSSkwi9StBXHf3LX031wg7OqldAGeGvKOnAylr2sv5B3vSo4zn1cA+zChEdf84HWoeI9j2tA4/mfDee8oSHIaSC3aYQ4tnNISRSwEudbU112u/ddnixJZ73koc8SMlOPsNTnvRyp9njKldQvRWIpw5iu861L3oKUV7njkdb8nk3fsyT3nBF9NzjNsp7z7XPQ4Lr2/am57vgxVRzV8db5lqKfOPFrkJOh93pORd5seWvd5Jb3EHVlrWdWi1xP6tf2HqWPJxr8EBquxDPxqe82wlfaeu7YVVttsMgDrGIR0ziEpv4IgEBADs="
        }
     ]
  },
}
```

**Reply by rockgobo** [reply]
Like the idea of having a schema and the metadataobject "_"
